File: runners/s3-benchrunner-java/src/main/java/com/example/s3benchrunner/BenchmarkRunner.java
package com.example.s3benchrunner;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.concurrent.Semaphore;

public abstract class BenchmarkRunner {

    public BenchmarkConfig config;
    public String bucket;
    public String region;

    public byte[] randomDataForUpload;
    private static final int MAX_CONCURRENCE = 1000;
    public Semaphore concurrency_semaphore = new Semaphore(MAX_CONCURRENCE, true);

    public BenchmarkRunner(BenchmarkConfig config, String bucket, String region) {
        this.config = config;
        this.bucket = bucket;
        this.region = region;
        if (!config.filesOnDisk) {
            this.randomDataForUpload = Util.generateRandomData();
        }
    }

    abstract public void run();

    public void prepareRun() throws Exception {
        /* Preparation work between runs */
        for (var task : config.tasks) {
            if (task.action.equals("download")) {
                if (!Path.of(task.key).getParent().toFile().exists()) {
                    Files.createDirectories(Path.of(task.key).getParent());
                }
                if (Files.exists(Path.of(task.key))) {
                    Files.delete(Path.of(task.key));
                }
            }
        }
    };
}

File: runners/s3-benchrunner-dotnet/S3BenchRunner/BenchmarkRunner.cs
using S3BenchRunner.Models;

namespace S3BenchRunner;

public abstract class BenchmarkRunner
{
    protected WorkloadConfig Config { get; }
    protected string Bucket { get; }
    protected string Region { get; }

    protected BenchmarkRunner(WorkloadConfig config, string bucket, string region)
    {
        Config = config;
        Bucket = bucket;
        Region = region;
    }

    public abstract Task RunAsync();

    public void PrepareRun()
    {
        // Preparation work between runs
        foreach (var task in Config.Tasks)
        {
            if (task.Action == "download")
            {
                var path = task.LocalPath;
                var directory = Path.GetDirectoryName(path);
                
                // Create directory if it doesn't exist
                if (!string.IsNullOrEmpty(directory) && !Directory.Exists(directory))
                {
                    Directory.CreateDirectory(directory);
                }
                
                // Delete file if it exists
                if (File.Exists(path))
                {
                    File.Delete(path);
                }
            }
        }
    }
}